### PR TITLE
Fix bug in `str_trunc()` with simpler code

### LINKED ISSUE #512: str_trunc() incorrectly snips rhs-of-ellipsis for truncated strings
**str_trunc** with param `side = 'left'` or `side = 'center'` returns incorrect result when ellipsis is the the same size or one larger than `width`. The rhs of the `ellipsis` gets appended without trunctation. 

``` r
datr <- c('', 'a', 'aa', 'aaa', 'aaaa', 'aaaaaaa')

stringr::str_trunc(datr, width=4, side = 'left',   ellipsis = "..") # correct
#> [1] ""     "a"    "aa"   "aaa"  "aaaa" "..aa"
stringr::str_trunc(datr, width=4, side = 'right',  ellipsis = "..") # correct
#> [1] ""     "a"    "aa"   "aaa"  "aaaa" "aa.."
stringr::str_trunc(datr, width=4, side = 'center', ellipsis = "..") # correct
#> [1] ""     "a"    "aa"   "aaa"  "aaaa" "a..a"
stringr::str_trunc(datr, width=3, side = 'left',   ellipsis = "..") # correct
#> [1] ""    "a"   "aa"  "aaa" "..a" "..a"
stringr::str_trunc(datr, width=3, side = 'right',  ellipsis = "..") # correct
#> [1] ""    "a"   "aa"  "aaa" "a.." "a.."
stringr::str_trunc(datr, width=3, side = 'center', ellipsis = "..") # should be c("","a","aa","aaa","a..","a..")
#> [1] ""           "a"          "aa"         "aaa"        "a..aaaa"      "a..aaaaaaa"
stringr::str_trunc(datr, width=2, side = 'left',   ellipsis = "..") # should be c("","a","aa","..","..","..")
#> [1] ""          "a"         "aa"        "..aaa"     "..aaaa"    "..aaaaaaa"
stringr::str_trunc(datr, width=2, side = 'right',  ellipsis = "..") # correct
#> [1] ""   "a"  "aa" ".." ".." ".."
stringr::str_trunc(datr, width=2, side = 'center', ellipsis = "..") # should be c("","a","aa","..","..","..")
#> [1] ""          "a"         "aa"        "..aaa"     "..aaaa"    "..aaaaaaa"
stringr::str_trunc(datr, width=1, side = 'left',   ellipsis = "..") # correct
#> Error in `stringr::str_trunc()`:
stringr::str_trunc(datr, width=1, side = 'right',  ellipsis = "..") # correct
#> Error in `stringr::str_trunc()`:
stringr::str_trunc(datr, width=1, side = 'center', ellipsis = "..") # correct
#> Error in `stringr::str_trunc()`:

stringr::str_trunc(datr, width=4, side = 'left',   ellipsis = "~") # correct
#> [1] ""     "a"    "aa"   "aaa"  "aaaa" "~aaa"
stringr::str_trunc(datr, width=4, side = 'right',  ellipsis = "~") # correct
#> [1] ""     "a"    "aa"   "aaa"  "aaaa" "aaa~"
stringr::str_trunc(datr, width=4, side = 'center', ellipsis = "~") # correct
#> [1] ""     "a"    "aa"   "aaa"  "aaaa" "aa~a"
stringr::str_trunc(datr, width=3, side = 'left',   ellipsis = "~") # correct
#> [1] ""    "a"   "aa"  "aaa" "~aa" "~aa"
stringr::str_trunc(datr, width=3, side = 'right',  ellipsis = "~") # correct
#> [1] ""    "a"   "aa"  "aaa" "aa~" "aa~"
stringr::str_trunc(datr, width=3, side = 'center', ellipsis = "~") # correct
#> [1] ""    "a"   "aa"  "aaa" "a~a" "a~a"
stringr::str_trunc(datr, width=2, side = 'left',   ellipsis = "~") # correct
#> [1] ""   "a"  "aa" "~a" "~a" "~a"
stringr::str_trunc(datr, width=2, side = 'right',  ellipsis = "~") # correct
#> [1] ""   "a"  "aa" "a~" "a~" "a~"
stringr::str_trunc(datr, width=2, side = 'center', ellipsis = "~") # should be c("","a","aa","a~","a~","a~")
#> [1] ""          "a"         "aa"        "a~aaa"     "a~aaaa"    "a~aaaaaaa"
stringr::str_trunc(datr, width=1, side = 'left',   ellipsis = "~") # should be c("","a","~","~","~","~")
#> [1] ""         "a"        "~aa"      "~aaa"     "~aaaa"    "~aaaaaaa"
stringr::str_trunc(datr, width=1, side = 'right',  ellipsis = "~") # correct
#> [1] ""  "a" "~" "~" "~" "~"
stringr::str_trunc(datr, width=1, side = 'center', ellipsis = "~") # should be c("","a","~","~","~","~")
#> [1] ""         "a"        "~aa"      "~aaa"     "~aaaa"    "~aaaaaaa"
stringr::str_trunc(datr, width=0, side = 'left',   ellipsis = "~") # correct
#> Error in `stringr::str_trunc()`:
stringr::str_trunc(datr, width=0, side = 'right',  ellipsis = "~") # correct
#> Error in `stringr::str_trunc()`:
stringr::str_trunc(datr, width=0, side = 'center', ellipsis = "~") # correct
#> Error in `stringr::str_trunc()`:

stringr::str_trunc(datr, width=3, side = 'left',   ellipsis = "") # correct
#> [1] ""    "a"   "aa"  "aaa" "aaa" "aaa"
stringr::str_trunc(datr, width=3, side = 'right',  ellipsis = "") # correct 
#> [1] ""    "a"   "aa"  "aaa" "aaa" "aaa"
stringr::str_trunc(datr, width=3, side = 'center', ellipsis = "") # correct 
#> [1] ""    "a"   "aa"  "aaa" "aaa" "aaa"
stringr::str_trunc(datr, width=2, side = 'left',   ellipsis = "") # correct
#> [1] ""   "a"  "aa" "aa" "aa" "aa"
stringr::str_trunc(datr, width=2, side = 'right',  ellipsis = "") # correct 
#> [1] ""   "a"  "aa" "aa" "aa" "aa"
stringr::str_trunc(datr, width=2, side = 'center', ellipsis = "") # correct 
#> [1] ""   "a"  "aa" "aa" "aa" "aa"
stringr::str_trunc(datr, width=1, side = 'left',   ellipsis = "") # correct
#> [1] ""  "a" "a" "a" "a" "a"
stringr::str_trunc(datr, width=1, side = 'right',  ellipsis = "") # correct 
#> [1] ""  "a" "a" "a" "a" "a"
stringr::str_trunc(datr, width=1, side = 'center', ellipsis = "") # should be c("","a","a","a","a","a")
#> [1] ""         "a"        "aaa"      "aaaa"     "aaaaa"    "aaaaaaaa"
stringr::str_trunc(datr, width=0, side = 'left',   ellipsis = "") # should be c("","","","","","")
#> [1] ""        "a"       "aa"      "aaa"     "aaaa"    "aaaaaaa"
stringr::str_trunc(datr, width=0, side = 'right',  ellipsis = "") # correct
#> [1] "" "" "" "" "" ""
stringr::str_trunc(datr, width=0, side = 'center', ellipsis = "") # should be c("","","","","","")
#> [1] ""        "a"       "aa"      "aaa"     "aaaa"    "aaaaaaa"
```

<sup>Created on 2023-06-03 with [reprex v2.0.2.9000](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by UchidaMizuki**
Fix #512 .
This is a pull request similar to #513 .

#512 is caused by the boundary cases of `stringi::stri_sub()` when getting the end of a string.
https://github.com/gagolews/stringi/issues/494

I modified the code to avoid the boundary cases, i.e., `from` is always a positive number, and added tests.
I think this proposal can fix the bug with simpler code than #513 .

### CODE REVIEW

**Comment by eauleaf** [comment]
Yeah, this fix is cleaner. Looked at speed; they're equivalent.

**Review by hadley — APPROVED** [review]
Thanks for working on this! Can you please add a bullet to the top of `NEWS.md`? It should briefly [describe the change](http://style.tidyverse.org/news.html) and end with `(@yourname, #issuenumber)`.

**Comment by UchidaMizuki** [comment]
I have updated NEWS.md. Please check it.

File: NEWS.md [hunk 1]
```diff
@@ -4,6 +4,9 @@
 
 ## Breaking changes
 
+* `str_trunc()` now correctly truncates strings when `side` is `"left"` or
```
Comment (hadley) [inline NEWS.md:7]: Can you please put this in the news for the development version?

Comment (UchidaMizuki) [reply NEWS.md]: Sorry for the mistake. I just fixed it.

**Comment by hadley** [comment]
Thanks!
